static/app/utils/useAsync.tsx

```diff
@@ -0,0 +1,382 @@
+import React, {useEffect, useState} from 'react';
+import * as Sentry from '@sentry/react';
+
+import {ResponseMeta} from 'sentry/api';
+import LoadingError from 'sentry/components/loadingError';
+import LoadingIndicator from 'sentry/components/loadingIndicator';
+import {t} from 'sentry/locale';
+import {metric} from 'sentry/utils/analytics';
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import useLocation from 'sentry/utils/useLocation';
+import useParams from 'sentry/utils/useParams';
+import useRoutes from 'sentry/utils/useRoutes';
+import PermissionDenied from 'sentry/views/permissionDenied';
+import RouteError from 'sentry/views/routeError';
+
+import RequestError from './requestError/requestError';
+
+type State = {
+  data: {[key: string]: any};
+  errors: {[key: string]: RequestError};
+  hasError: boolean;
+  isLoading: boolean;
+  isReloading: boolean;
+  remainingRequests: number;
+};
```

Comment:
It would be really nice if we could type this better. 

Right now `data` is just anything, it would be cool if we could templatize this so when you're invoking `useAsync` we can specify what the data types will look like